File: xssfinder-executor-java/src/test/java/org/xssfinder/remote/ExecutorHandlerTest.java
package org.xssfinder.remote;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.xssfinder.runner.ExecutorContext;
import org.xssfinder.runner.TraversalResult;
import org.xssfinder.scanner.PageDefinitionFactory;
import org.xssfinder.scanner.PageFinder;
import org.xssfinder.scanner.ThriftToReflectionLookup;
import org.xssfinder.scanner.ThriftToReflectionLookupFactory;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class ExecutorHandlerTest {
    private static final String PACKAGE_NAME = "some.namespace";
    private static final String PAGE_ID = "Page ID";

    @Mock
    private PageFinder mockPageFinder;
    @Mock
    private PageDefinitionFactory mockPageDefFactory;
    @Mock
    private ThriftToReflectionLookupFactory mockThriftToReflectionLookupFactory;
    @Mock
    private ExecutorContext mockExecutorContext;
    @Mock
    private PageDefinition mockPageDefinition;
    @Mock
    private ThriftToReflectionLookup mockLookup = new ThriftToReflectionLookup();

    @InjectMocks
    private ExecutorHandler executorHandler;

    @Before
    public void setUp() {
        when(mockThriftToReflectionLookupFactory.createLookup()).thenReturn(mockLookup);
    }

    @Test
    public void gettingPageDefinitionsDelegatesToPageFinderAndPageDefinitionFactory() throws Exception {
        // given
        Set<Class<?>> pageClasses = new HashSet<Class<?>>();
        pageClasses.add(SomePage.class);
        when(mockPageFinder.findAllPages(PACKAGE_NAME)).thenReturn(pageClasses);
        when(mockPageDefFactory.createPageDefinition(SomePage.class, pageClasses, mockLookup)).thenReturn(mockPageDefinition);

        // when
        Set<PageDefinition> pageDefinitions = executorHandler.getPageDefinitions(PACKAGE_NAME);

        // then
        Set<PageDefinition> expectedDefinitions = ImmutableSet.of(mockPageDefinition);
        assertThat(pageDefinitions, is(expectedDefinitions));
    }

    @Test
    public void gettingPageDefinitionsSetsContextLookup() throws Exception {
        // when
        executorHandler.getPageDefinitions(PACKAGE_NAME);

        // then
        verify(mockExecutorContext).setThriftToReflectionLookup(mockLookup);
    }

    @Test
    public void gettingPageDefinitionsThrowsExceptionIfNoPageObjectsCanBeFound() throws Exception {
        // given
        Set<Class<?>> pageClasses = new HashSet<Class<?>>();
        when(mockPageFinder.findAllPages(PACKAGE_NAME)).thenReturn(pageClasses);

        // when
        executorHandler.getPageDefinitions(PACKAGE_NAME);
    }

    @Test
    public void visitingUrlIsDelegatedToExecutorContext() throws Exception {
        // when
        executorHandler.startRoute("pageId");

        // then
        verify(mockExecutorContext).visitUrlOfRootPage("pageId");
    }

    @Test
    public void puttingXssAttackStringsInInputsIsDelegatedToExecutorContext() throws Exception {
        // given
        Map<String, String> givenInputIdsToAttackIds = ImmutableMap.of("foo", "bar");
        when(mockExecutorContext.putXssAttackStringsInInputs()).thenReturn(givenInputIdsToAttackIds);

        // when
        Map<String, String> inputIdsToAttackIds = executorHandler.putXssAttackStringsInInputs();

        // then
        assertThat(inputIdsToAttackIds, is(givenInputIdsToAttackIds));
    }

    @Test
    public void gettingCurrentXssIdsIsDelegatedToExecutorContext() throws Exception {
        // given
        Set<String> givenIds = ImmutableSet.of("foo");
        when(mockExecutorContext.getCurrentXssIds()).thenReturn(givenIds);

        // when
        Set<String> xssIds = executorHandler.getCurrentXssIds();

        // then
        assertThat(xssIds, is(givenIds));
    }

    @Test
    public void gettingFormCountIsDelegatedToExecutorContext() throws Exception {
        // given
        int givenFormCount = 3;
        when(mockExecutorContext.getFormCount()).thenReturn(givenFormCount);

        // when
        int formCount =  executorHandler.getFormCount();

        // then
        assertThat(formCount, is(givenFormCount));
    }

    @Test
    public void traversingMethodIsDelegatedToExecutorContext() throws Exception {
        // given
        MethodDefinition mockMethodDefinition = mock(MethodDefinition.class);
        TraversalResult mockResult = mock(TraversalResult.class);
        Map<String, String> expectedInputIdsToXssIds = new HashMap<String, String>();
        when(mockResult.getInputIdsToAttackIds()).thenReturn(expectedInputIdsToXssIds);
        when(mockExecutorContext.traverseMethod(mockMethodDefinition, TraversalMode.NORMAL)).thenReturn(mockResult);

        // when
        Map<String, String> inputIdsToXssIds = executorHandler.traverseMethod(mockMethodDefinition, TraversalMode.NORMAL);

        // then
        assertThat(inputIdsToXssIds, is(expectedInputIdsToXssIds));
    }

    @Test
    public void invokingAfterRouteHandlerDelegatesToExecutorContext() throws Exception {
        // when
        executorHandler.invokeAfterRouteHandler(PAGE_ID);

        // then
        verify(mockExecutorContext).invokeAfterRouteHandler(PAGE_ID);
    }

    private static class SomePage {
        @SuppressWarnings("UnusedDeclaration")
        public SomePage goToSomePage() { return null; }
    }
}

File: xssfinder-php-executor/test/XssFinder/Remote/ExecutorHandlerTest.php
<?php

namespace XssFinder\Remote;

use PHPUnit_Framework_TestCase;
use XssFinder\MethodDefinition;
use XssFinder\PageDefinition;
use XssFinder\Remote\ExecutorHandler;
use XssFinder\Runner\ExecutorContext;
use XssFinder\Runner\TraversalResult;
use XssFinder\Scanner\PageDefinitionFactory;
use XssFinder\Scanner\PageFinder;
use XssFinder\Scanner\PageFinderFactory;
use XssFinder\Scanner\ThriftToReflectionLookup;
use XssFinder\Scanner\ThriftToReflectionLookupFactory;
use XssFinder\TraversalMode;

class ExecutorHandlerTest extends PHPUnit_Framework_TestCase
{
    const SOME_NAMESPACE = 'namespace';

    /** @var PageFinderFactory */
    private $_mockPageFinderFactory;
    /** @var PageFinder */
    private $_mockPageFinder;
    /** @var PageDefinitionFactory */
    private $_mockPageDefinitionFactory;
    /** @var ThriftToReflectionLookupFactory */
    private $_mockLookupFactory;
    /** @var ExecutorContext */
    private $_mockExecutorContext;

    /** @var ThriftToReflectionLookup */
    private $_mockLookup;

    /** @var ExecutorHandler */
    private $_handler;

    public function setUp()
    {
        $this->_mockPageFinderFactory = mock('\XssFinder\Scanner\PageFinderFactory');
        $this->_mockPageFinder = mock('\XssFinder\Scanner\PageFinder');
        $this->_mockPageDefinitionFactory = mock('\XssFinder\Scanner\PageDefinitionFactory');
        $this->_mockLookupFactory = mock('\XssFinder\Scanner\ThriftToReflectionLookupFactory');
        $this->_mockExecutorContext = mock('\XssFinder\Runner\ExecutorContext');

        $this->_mockLookup = mock('\XssFinder\Scanner\ThriftToReflectionLookup');
        when($this->_mockLookupFactory->createLookup())->return($this->_mockLookup);

        when($this->_mockPageFinderFactory->createPageFinder(self::SOME_NAMESPACE))->return($this->_mockPageFinder);

    }

    public function testGettingPageDefinitionsIsDelegatedToFinderAndFactoryAndLookupIsSetOnContext()
    {
        // given
        $classNames = array('SomePage', 'NotAPage');
        $pageNames = array('SomePage');
        when($this->_mockPageFinder->findPages($classNames))->return($pageNames);
        /** @var PageDefinition $mockPageDefinition */
        $mockPageDefinition = mock('XssFinder\PageDefinition');
        when($this->_mockPageDefinitionFactory->createPageDefinition(
            'SomePage', argOfTypeThat('XssFinder\Scanner\ThriftToReflectionLookup', anything()))
        )->return($mockPageDefinition);
        $this->_handler = $this->_createExecutorHandler($classNames);

        // when
        $pageDefinitions = $this->_handler->getPageDefinitions(self::SOME_NAMESPACE);

        // then
        assertThat($pageDefinitions, equalTo(array($mockPageDefinition)));
        verify($this->_mockExecutorContext)->setThriftToReflectionLookup($this->_mockLookup);
    }

    public function testStartingRouteIsDelegatedToExecutorContext()
    {
        // given
        $this->_handler = $this->_createExecutorHandler();

        // when
        $this->_handler->startRoute('SomePage');

        // then
        verify($this->_mockExecutorContext)->visitUrlOfRootPage('SomePage');
    }

    public function testPuttingXssAttackStringsInInputsIsDelegatedToExecutorContext()
    {
        // given
        $this->_handler = $this->_createExecutorHandler();
        $expectedResult = array('inputId' => 'attackId');
        when($this->_mockExecutorContext->putXssAttackStringsInInputs())->return($expectedResult);

        // when
        $inputIdsToAttackIds = $this->_handler->putXssAttackStringsInInputs();

        // then
        assertThat($inputIdsToAttackIds, is($expectedResult));
    }

    public function testTraversingMethodIsDelegatedToExecutorContext()
    {
        // given
        $this->_handler = $this->_createExecutorHandler();
        /** @var MethodDefinition $mockMethodDefinition */
        $mockMethodDefinition = mock('XssFinder\MethodDefinition');
        /** @var TraversalResult $mockTraversalResult */
        $mockTraversalResult = mock('XssFinder\Runner\TraversalResult');
        $expectedInputIdsToXssIds = array('//some/input' => '123');
        when($mockTraversalResult->getInputIdsToAttackIds())->return($expectedInputIdsToXssIds);
        when($this->_mockExecutorContext->traverseMethod($mockMethodDefinition, TraversalMode::NORMAL))
            ->return($mockTraversalResult);

        // when
        $inputIdsToXssIds = $this->_handler->traverseMethod($mockMethodDefinition, TraversalMode::NORMAL);

        // then
        assertThat($inputIdsToXssIds, is($expectedInputIdsToXssIds));
    }

    public function testGettingCurrentXssIdsIsDelegatedToExecutorContext()
    {
        // given
        $this->_handler = $this->_createExecutorHandler();
        when($this->_mockExecutorContext->getCurrentXssIds())->return(array('1','2'));

        // when
        $currentXssIds = $this->_handler->getCurrentXssIds();

        // then
        assertThat($currentXssIds, is(array('1','2')));
    }

    public function testGettingFormCountIsDelegatedToExecutorContext()
    {
        // given
        $this->_handler = $this->_createExecutorHandler();
        when($this->_mockExecutorContext->getFormCount())->return(3);

        // when
        $formCount = $this->_handler->getFormCount();

        // then
        assertThat($formCount, is(3));
    }

    private function _createExecutorHandler($classNames = array())
    {
        return new ExecutorHandler(
            $this->_mockPageFinderFactory,
            $this->_mockPageDefinitionFactory,
            $this->_mockLookupFactory,
            $this->_mockExecutorContext,
            $classNames
        );
    }
}